s3migrate/shared.py

```diff
@@ -0,0 +1,397 @@
+                md5 = md5_fileobj_b64(tmp)
+                tmp.seek(0)
+                target = s3_resource.Object(
+                    s3_bucket,
+                    self.make_contentrepo_key())
+                response = target.put(
+                    Body=tmp,
+                    ContentMD5=md5)
+                return response["ETag"].replace("\"", "")
+
+    def migrate(self, mogile_client, table, s3_resource, bucket_map):
+        """Migrate this mogile object to contentrepo.
+
+        Returns None if the object is a temporary file, otherwise
+        returns the md5 of the migrated file.
+        """
+        if self.temp is True:
+            return None  # Do not migrate temporary files.
+        s3_bucket = bucket_map[self.mogile_bucket]
+        print(f"Migrating {self.fid} to "
+              f"s3://{s3_bucket}/{self.make_contentrepo_key()}")
+        md5 = self.contentrepo_exists_in_bucket(s3_resource, s3_bucket)
+        try:
+            if md5 is not False:
+                # Migration done!
+                return True
+            if self.intermediary_exists_in_bucket(s3_resource, s3_bucket):
+                print(f"  Copying from s3://{s3_bucket}/"
+                      f"{self.make_intermediary_key()}")
+                md5 = self.copy_from_intermediary(s3_resource, s3_bucket)
+                return True
+            # Nothing is on S3 yet, copy content directly to the
+            # final location.
+            print(f"  Putting from mogile.")
+            md5 = self.put(mogile_client, s3_resource, s3_bucket)
+        finally:
+            if md5 is not False:
+                self.save_to_dynamodb(table, md5, s3_bucket)
+
+    def save_to_dynamodb(self, table, md5, s3_bucket):
+        """Save record to dynamodb certifying successful migration."""
+        return table.put_item(
+            Item={
+                'fid': self.fid,
+                'sha1': self.sha1sum,
+                'md5': md5,
+                'bucket': s3_bucket
+            })
+
+    def to_json(self):
+        """Serialize as JSON."""
+        return json.dumps({
+            "length": self.length,
+            "fid": self.fid,
+            "dkey": self.dkey
+            })
+
+    @classmethod
+    def from_json(cls, json_str):
+        """Create MogileFile from JSON string."""
+        return MogileFile(**json.loads(json_str))
+
+    def verify(self, fid, md5, sha1, bucket, bucket_map, s3_resource):
+        """Verify (with assert) this mogile file against asserted values."""
+        print(f"Verifying {self.fid}")
+        assert self.fid == fid, \
+            f"fid {self.fid} not migrated"
+        new_bucket = bucket_map[self.mogile_bucket]
+        assert new_bucket == bucket
+        s3_md5 = self.contentrepo_exists_in_bucket(s3_resource, new_bucket)
+        assert (s3_md5 == md5), f"{self.fid} has wrong MD5 sum"
+        assert (self.sha1sum == sha1), \
+            f"{self.fid} has wrong SHA1 sum"
```

Comment:
md5 check aaaand sha check. Sorry since I'm sure you've explained this before, but why do we need both?

Reply:
On the one hand, it's kind of belt and suspenders double-checking.

But mostly it's because AWS reports the md5 sum via etag, while contentrepo uses the sha1 sum for the content-addressable storage. We check the reported value from the etag in S3 while we just verify the sha1 sum stored in dynamodb.

Reply:
gotcha